config/schedule.yml

```diff
@@ -0,0 +1,3 @@
+cancel_reservation_job:
+  cron: '*/5 * * * *'
+  class: 'CancelOutdatedReservationsJob'
```

Comment:
Why do you want to create a cron job instead of calling a job and scheduling it to 30 minutes? In this case, it could be a situation when the cron job was performed 1 minute ago and now I'm buying a ticket so it will be checked when the ticket was created 29 minutes ago and next time when it was created 34 minutes ago. The ticket should have been canceled for 4 minutes. 

Definitely a better solution is to schedule a job for 30minutes when a reservation has been created.

Reply:
Because I told them that better solution is to handle that as cron job 😄

Reasoning for that:
- Reservations should be canceled 30 minutes before the screening starts, not 30 minutes after booking
- This has pretty nasty consequences - if we have few screenings at the same time, it means there could be maybe oven over thousand reservations that each will have job queued for the same time, which could block sidekiq for some duration, as that 1000+ jobs need to execute. Because those will execute for all reservations, not only not paid ones. Cron job will run only one and handle all jobs
- it handles edge cases better - eg. if time of screening will change. Let's say that somebody bought ticket a week earlier for some big premiere. The ticket was for first screening in the morning, at 7:00 AM. But then two days later we will learn that Tauron will be doing something and we will not have electricity till 8 AM. So we are moving this screening to 9AM. Now cron job will handle this change without any issue, but job scheduled at time of reservation will not - it will still run at 6:30 AM.
- Cron jobs better handle errors - if for some reason there will be issue with DB for 10 minutes, scheduled job will exhaust all retries and fail never canceling the tickets. Cron job on the other hand will still cancel them, just 10 minutes later.
- yes, there are edge cases with screenings set for times like 7:32 - but that's unlikely to be actual screening time, as most cinemas try to set them for at least times divisible by 5. And it can be fixed by just running cron job each minutes if it would be needed.

Reply:
I think, there is no other way. If you allow your users to create reservations 30 minutes before show start it will lead to some nasty consequences: Someone will make a reservation and next minute it will be cancelled or they might even buy them if they are quick enough and app doesn't protect against that.

Reply:
Yeah, users can not create reservations when seanse is less than 30 minutes away, but employes should be able to :smile: